lib/puppet_pal.rb

```diff
@@ -23,26 +23,236 @@
+    # puppet language string. The given string must form a complete and valid expression/statement as an error is raised
+    # otherwise. That is, it is not possible to divide a compound expression by line and evaluate each line individually.
+    #
+    # @param puppet_code [String, nil] the puppet language code to evaluate, must be a complete expression/statement
+    # @param source_file [String, nil] an optional reference to a source (a file or symbolic name/location) - defaults to '<unknown>'
+    #   in error messages.
+    # @return [Object] what the `puppet_code` evaluates to
+    #
+    def evaluate_string(puppet_code, source_file = nil)
+      return nil if puppet_code.nil? || puppet_code == ''
+      unless puppet_code.is_a?(String)
+        raise ArgumentError, _("The argument 'puppet_code' must be a String, got %{type}") % { type: puppet_code.class }
+      end
+      internal_evaluator.evaluate_string(topscope, puppet_code, source_file)
+    end
+
+    # Evaluates a puppet language file in top scope.
+    # The file must exist and contain valid puppet language code or an error is raised.
+    #
+    # @param file [Path, String] an absolute path to a file with puppet language code, must exist
+    # @return [Object] what the last evaluated expression in the file evaluated to
+    #
+    def evaluate_file(file)
+      internal_evaluator.evaluate_file(topscope, file)
+    end
+
+    # Parses a puppet data type given in String format and returns that type, or raises an error.
+    # A type is needed in calls to `new` to create an instance of the data type, or to perform type checking
+    # of values - typically using `type.instance?(obj)` to check if `obj` is an instance of the type.
+    #
+    # @example Verify if obj is an instance of a data type
+    #   # evaluates to true
+    #   pal.type('Enum[red, blue]').instance?("blue")
+    #
+    # @example Create an instance of a data type
+    #   # using an already create type
+    #   t = pal.type('Car')
+    #   pal.new_object(t, [{'color' => 'black', 'make' => 't-ford}])
+    #
+    #   # letting 'new_object' parse the type from a string
+    #   pal.new_object('Car', [{'color' => 'black', 'make' => 't-ford}])
+    #
+    # @param type_string [String] a puppet language data type
+    # @return [Puppet::Pops::Types::TypedModelObject] the data type
+    #
+    def type(type_string)
+      Puppet::Pops::Types::TypeParser.singleton.parse(type_string)
+    end
+
+    # Creates a new instance of a given data type.
+    # @param data_type [String, Puppet::Pops::Types::PAnyType] the data type as a data type or in String form.
+    # @param arguments [Array] an array of the arguments to `new`, defaults to empty array.
+    # @return [Object] an instance of the given data type,
+    #   or raises an error if it was not possible to parse data type or create an instance.
+    #
+    def new_object(data_type, arguments = [])
```

Comment:
Why not use `*arguments` here? It will default to `[]` when no arguments are given. It will also remove the need to always wrap a hash of named arguments (as done in the examples above) and enable calls like `pal.new_object('Car', 'color' => 'black', 'make' => 't-ford')

Reply:
For consistency with call_function where args are also an array